NeosModLoader/Utility/GenericTypeMethodsInvoker.cs

```diff
@@ -0,0 +1,166 @@
+		{
+			GenericType = genericType;
+			GetGenericMethodOfConcreteType = getGenericMethodOfConcreteType;
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type[] instanceTypes, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceTypes, instance, methodTypes, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type[] instanceTypes, object instance, Type methodType, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceTypes, instance, methodType, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, instance, methodTypes, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type instanceType, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceType, instance, methodTypes, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, object instance, Type methodType, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, instance, methodType, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type instanceType, object instance, Type methodType, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceType, instance, methodType, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, instance, methodTypes, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, object instance, Type methodType, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, instance, methodType, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, object instance, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, instance, new TypeDefinition(), parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, null, new TypeDefinition(), parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, object instance, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, instance, new TypeDefinition(), parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, null, new TypeDefinition(), parameters);
+		}
+
+		private static MethodInfo GetGenericMethodOfConcreteTypeDefault(MethodInfo needleMethod, Type concreteType)
```

Comment:
There are a few things I'm confused about.

1. Wouldn't something like this be equivalent? `AccessTools.Method(needleMethod.Name, needleMethod.GetParameters())` Or is there something I'm not seeing?
2. Needle and Hay are some bonkers variable names. I had a tough time grokking all this code.
3. If this works consistently then why is there a constructor that lets users pass a custom `getGenericMethodOfConcreteType` implementation? If this doesn't work consistently then why have it at all?

Reply:
> 1. Wouldn't something like this be equivalent? `AccessTools.Method(needleMethod.Name, needleMethod.GetParameters())` Or is there something I'm not seeing?

I have no idea how that handles instances of the generic parameters of the containing type or the method 🤔 


> 2\. Needle and Hay are some bonkers variable names. I had a tough time grokking all this code.

Went with the needle and haystack metaphor for searching something, but I guess it's not super clear in this context lol


> 3\. If this works consistently then why is there a constructor that lets users pass a custom `getGenericMethodOfConcreteType` implementation? If this doesn't work consistently then why have it at all?

I'm not sure, having the Func is kind of just what I first came up with